File: CSE221-Lab-Codes/SugestedProb12.py
import sys
sys.setrecursionlimit(10**7)
input = sys.stdin.readline

def get_inorder(pre, post):
    if not pre:
        return []
    root = pre[0]
    if len(pre) == 1:
        return [root]

    left_root = pre[1]
    k = post.index(left_root) 

    left_size = k + 1

    left_pre = pre[1:1+left_size]
    right_pre = pre[1+left_size:]

    left_post = post[:left_size]
    right_post = post[left_size:-1]

    left = get_inorder(left_pre, left_post)
    right = get_inorder(right_pre, right_post)

    return left + [root] + right


n = int(input())

preorder = list(map(int, input().split()))

postorder = list(map(int, input().split()))

result = get_inorder(preorder, postorder)

print(*result)

File: Lab 3/Java/SugestedProb12.java
import java.util.*;

public class SugestedProb12 {
    static List<Integer> getInorder(List<Integer> pre, List<Integer> post) {
        if (pre.isEmpty())
            return new ArrayList<>();
        int root = pre.get(0);
        if (pre.size() == 1) {
            List<Integer> res = new ArrayList<>();
            res.add(root);
            return res;
        }
        int leftRoot = pre.get(1);
        int k = post.indexOf(leftRoot);
        int leftSize = k + 1;

        List<Integer> leftPre = pre.subList(1, 1 + leftSize);
        List<Integer> rightPre = pre.subList(1 + leftSize, pre.size());

        List<Integer> leftPost = post.subList(0, leftSize);
        List<Integer> rightPost = post.subList(leftSize, post.size() - 1);

        List<Integer> left = getInorder(leftPre, leftPost);
        List<Integer> right = getInorder(rightPre, rightPost);

        List<Integer> res = new ArrayList<>();
        res.addAll(left);
        res.add(root);
        res.addAll(right);
        return res;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt())
            return;
        int n = sc.nextInt();
        List<Integer> pre = new ArrayList<>();
        for (int i = 0; i < n; i++)
            pre.add(sc.nextInt());
        List<Integer> post = new ArrayList<>();
        for (int i = 0; i < n; i++)
            post.add(sc.nextInt());

        List<Integer> res = getInorder(pre, post);
        for (int val : res)
            System.out.print(val + " ");
        System.out.println();
    }
}
